LibGUI+SystemMonitor: Enable searching on the processes list

d1823 opened the pull request:
This enables the System Monitor's ProcessModel to be searched in the same manner as FileSystemModel and ItemListModel.
The corresponding LibGUI change handles a case where the highlighting would get out of sync right after the SystemMonitor's timer updated the list and order of processes.

alimpfard commented on Userland/Libraries/LibGUI/AbstractView.cpp:
```diff
@@ -72,6 +72,7 @@ void AbstractView::model_did_update(unsigned int flags)
         if (!model()->is_within_range(m_drop_candidate_index))
             m_drop_candidate_index = {};
         selection().remove_matching([this](auto& index) { return !model()->is_within_range(index); });
+        do_search(move(m_searching));
```
Since this seems exceptionally weird at first glance, could we add a `find_next_search_match()` next to `do_search()` that doesn't touch `m_searching` and just finds the next match?
Assuming I'm right about what this intends to do anyway

d1823 replied:
You're absolutely right. The `find_next_search_match()` is a good idea. Thanks!

d1823 commented:
You guys are right this highlighting part looks quite weird. I decided to go ahead and do a small refactor to make the whole thing a little bit more readable. I renamed a few variables and moved some stuff around. I also split the `AbstractView::do_search` into `AbstractView::find_next_search_match` and `AbstractView::highlight_search` while retaining the same behavior. Thanks to that I was able to tweak the `AbstratView::model_did_update` in a way that doesn't require moving that string around.

I kind of feel like I got this PR a bit offtrack so if you want that to be split into two separate PRs, I'll be okay with that.

That was quite a bit of fun. :smile:

kleinesfilmroellchen commented on Userland/Applications/SystemMonitor/ProcessModel.cpp:
```diff
@@ -313,6 +313,24 @@ GUI::Variant ProcessModel::data(const GUI::ModelIndex& index, GUI::ModelRole rol
+    Vector<GUI::ModelIndex, 1> found_indices;
+
+    for (auto& thread : m_threads) {
+        if (string_matches((*thread.value).current_state.name, searching, flags)) {
```
The pointer dereference seems unnecessary. Please use `->` instead.

kleinesfilmroellchen commented on Userland/Applications/SystemMonitor/ProcessModel.cpp:
```diff
@@ -313,6 +313,24 @@ GUI::Variant ProcessModel::data(const GUI::ModelIndex& index, GUI::ModelRole rol
     return {};
 }
 
+Vector<GUI::ModelIndex, 1> ProcessModel::matches(const StringView& searching, unsigned flags, const GUI::ModelIndex&)
```
Please don't specify an inline capacity unless you have a good reason to.

d1823 replied:
I understand and agree with the premise, but that pattern is literally in every other place related to in-table highlighted search. I believe it should be addressed (globally) with a separate PR, since this one is already big enough.

kleinesfilmroellchen replied:
I'd disagree that the PR is already big enough. (I myself have submitted single commits with significantly larger changes.) It's commonplace to refactor related code to the one you're working on, even if it's slightly unrelated. Just put a "Refactor XYZ" commit first, then your functional changes. After all, you're already renaming a bunch of members for clarity, which is a good thing.

kleinesfilmroellchen replied:
While we're here, east const please :^)

d1823 replied:
Huh. I guess one learns something new everyday. Sure then!

alimpfard reviewed (approved):
LGTM.

kleinesfilmroellchen commented on Userland/Applications/SystemMonitor/ProcessModel.h:
```diff
@@ -60,6 +60,8 @@ class ProcessModel final : public GUI::Model {
     virtual String column_name(int column) const override;
     virtual GUI::Variant data(const GUI::ModelIndex&, GUI::ModelRole) const override;
+    virtual bool is_searchable() const override { return true; }
+    virtual Vector<GUI::ModelIndex, 1> matches(const StringView&, unsigned = MatchesFlag::AllMatching, const GUI::ModelIndex& = GUI::ModelIndex()) override;
```
Let's use east const.
```suggestion
    virtual Vector<GUI::ModelIndex, 1> matches(StringView const&, unsigned = MatchesFlag::AllMatching, GUI::ModelIndex const& = GUI::ModelIndex()) override;
```

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/AbstractView.cpp:
```diff
@@ -663,12 +679,7 @@ void AbstractView::set_searchable(bool searchable)
-        cancel_searching();
-}
-
-bool AbstractView::is_highlighting_searching(const ModelIndex& index) const
```
More east const. If you're working on this file, you might as well improve all of them.

kleinesfilmroellchen reviewed (changes requested):
I think it's a good idea to pack up some (small) LibGUI refactoring with your commit to keep the code modern. After all, nobody can complain about a hygienic refactor :^)

d1823 commented:
Refactored both issues. This PR should be ready to merge at this point.
